### PR TITLE
[PORT] Ports asset_loading Subsystem + some asset related PRs + my vending machine fix

### PR DESCRIPTION
**Pull request by EvilDragonfiend**
## About The Pull Request

Partial & full ports:
* asset_loading.dm: https://github.com/tgstation/tgstation/commits/master/code/controllers/subsystem/asset_loading.dm
  * https://github.com/tgstation/tgstation/pull/69454
  * https://github.com/tgstation/tgstation/pull/70753
* asset files:
  * https://github.com/tgstation/tgstation/pull/66317
  * https://github.com/tgstation/tgstation/pull/63503
  * https://github.com/tgstation/tgstation/pull/69852
* vendor sprite fix (it showed glass icons wrongly)

Reverts
- https://github.com/BeeStation/BeeStation-Hornet/pull/6748

It was causing problems with asset sending and is inconsistent with TG who does not seem to have this issue. It also doesn't seem to happen anymore.

## Why It's Good For The Game
It's Lemon guys PR. can't be any bad.
also, had to touch a lot of codes to support the subsystem PR

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

image

image

image

image

image

image

image

image

</details>

## Changelog
:cl: LemonInTheDark, MrStonedOne, Mothblocks, EvilDragonfiend
add: Added cross-round serverside spritesheet caching, and various fixes and improvements for it.
config: Added CACHE_ASSETS, which enables asset caching when enabled. It is disabled in development, but should be enabled n production.
fix: Fixed vending machine showing icons as glass when they failed to load.
/:cl:

### CODE REVIEW

File: code/modules/admin/player_panel.dm [hunk 1]
```diff
@@ -308,11 +308,10 @@
 /datum/asset/spritesheet/antag_hud
 	name = "antag-hud"
```
Comment (itsmeow) [inline code/modules/admin/player_panel.dm:309]: this is cross round cachable

File: tgui/packages/tgui/interfaces/Vending.js [hunk 2]
```diff
@@ -134,7 +131,7 @@ export const Vending = (props, context) => {
                 key={product.name}
                 custom={custom}
                 product={product}
-                productStock={data.stock[product.path]} />
+                productStock={data.stock[product.name].amount} />
```
Comment (itsmeow) [inline tgui/packages/tgui/interfaces/Vending.js]: I *explicitly* changed this to use path and not name.
https://github.com/BeeStation/BeeStation-Hornet/pull/7229

Do not undo this

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/Vending.js (hunk 2 above)
Comment (EvilDragonfiend) [reply tgui/packages/tgui/interfaces/Vending.js]: fixed

**Review by PowerfulBacon — DISMISSED** [review]

**Review by itsmeow — DISMISSED** [review]
Looks fine. If something breaks we can just disable it via config.

**Comment by itsmeow** [comment]
@Crossedfall

**Comment by EvilDragonfiend** [comment]
Merge conflicted by PigeonVerde's sprite asset, and solved. no local issue found.

**Comment by EvilDragonfiend** [comment]
ah, it has more conflicts. I will fix it soon.

**Comment by EvilDragonfiend** [comment]
image

fixed

**Review by itsmeow — DISMISSED** [review]

File: code/controllers/configuration/entries/general.dm [hunk 3]
```diff
@@ -600,3 +600,6 @@
 /datum/config_entry/string/elasticsearch_metrics_apikey
 
 /datum/config_entry/flag/enable_mrat
+
+/datum/config_entry/flag/cache_assets
+	config_entry_value = TRUE
```
Comment (Crossedfall) [inline code/controllers/configuration/entries/general.dm]: Let's move this into `code\controllers\configuration\entries\resources.dm` to stay consistent

Comment (Crossedfall) [reply code/controllers/configuration/entries/general.dm]: Given how unintuitive it is to enable a feature by commenting it out in the config, let's default this to off instead and treat it like the other flags.

File: config/config.txt [hunk 4]
```diff
@@ -615,3 +615,10 @@ ELASTICSEARCH_METRICS_APIKEY thisIsSomethingThatsBased64Encoded==
 
 ## Enable/disable the mentor mrat verb ("Request Mentor Assistance")
 ENABLE_MRAT
+
+## Assets can opt-in to caching their results into `tmp`.
+## This is important, as preferences assets take upwards of 30 seconds (without sleeps) to collect.
+## The cache is assumed to be cleared by TGS recompiling, which deletes `tmp`.
+## This should be disabled (through `CACHE_ASSETS 0`) on development,
+## but enabled on production (the default).
+CACHE_ASSETS 0
```
Comment (Crossedfall) [inline config/config.txt]: Let's move this into `config/resources.txt` for consistency. Also be sure to update the language so that it clearly states that the option is disabled by default and can be enabled by uncommenting it.

**Review by Crossedfall — CHANGES_REQUESTED** [review]
Biggest issue I have now is just that the config entry is fairly unintuitive as is. Changing it so it's off by default would make it a bit more sensible.

I also think that we should move it into the resources config file just to improve the overall organization of things.

File: config/resources.txt [hunk 5]
```diff
@@ -37,3 +37,9 @@ ASSET_CDN_WEBROOT data/asset-store/
 # if you want to test this locally, you simpily run the `localhost-asset-webroot-server.py` python3 script to host assets stored in `data/asset-store/` via http://localhost:7123/
 ASSET_CDN_URL http://localhost:7123/
 
+## Assets can opt-in to caching their results into `tmp`.
+## This is important, as preferences assets take upwards of 30 seconds (without sleeps) to collect.
+## The cache is assumed to be cleared by TGS recompiling, which deletes `tmp`.
+## This should be disabled (through `CACHE_ASSETS 0`) on development
+## This should be enalbed by 1 from server side
+CACHE_ASSETS 0
```
Comment (Crossedfall) [inline config/resources.txt]:
```suggestion
########################
# Asset Caching Config #
########################
## Assets can opt-in to caching their results into `tmp`.
## This is important, as preferences assets take upwards of 30 seconds (without sleeps) to collect.
## The cache is assumed to be cleared by TGS recompiling, which deletes `tmp`.
## This should be uncommented on production
#CACHE_ASSETS
```

You don't actually need the whole `0` for disable and `1` for enabled now. You just comment it out to disable it and uncomment it to enable it.

**Comment by Crossedfall** [comment]
Alright, I'm reasonably happy with that. Going to do a TM and see how asset loading is actually impacted when it's properly enabled

**Comment by EvilDragonfiend** [comment]
player panel is fixed, but that might need a manual removal of asset folder

**Comment by itsmeow** [comment]
The changes you made are entirely cosmetic, I am not sure how it would fail with a prefix assigned.

**Comment by EvilDragonfiend** [comment]
> The changes you made are entirely cosmetic, I am not sure how it would fail with a prefix assigned.

tbh I don't know too but it fixed it lmao

**Comment by EvilDragonfiend** [comment]
fixed

**Comment by EvilDragonfiend** [comment]
checked the chem master

File: code/modules/asset_cache/asset_list.dm [hunk 6]
```diff
@@ -6,36 +7,57 @@
...
 
 /datum/asset/proc/register()
 	return
 
 /datum/asset/proc/send(client)
-	SHOULD_CALL_PARENT(TRUE)
-	if(!client)
-		return FALSE
-	return TRUE
+	return
```
Comment (itsmeow) [inline code/modules/asset_cache/asset_list.dm:56]: This needs to return true. new_player depends on this for opening the new player UI

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by EvilDragonfiend** [comment]
issue addressed.

**Comment by EvilDragonfiend** [comment]
lobby menu doesn't go away now

**Comment by itsmeow** [comment]
Do
- https://github.com/tgstation/tgstation/pull/66317
- https://github.com/tgstation/tgstation/pull/69852

apply at all?

**Comment by EvilDragonfiend** [comment]
> Do
> 
> * [[high-priority] Fix things breaking in tgui if the asset transport type is changed mid round tgstation/tgstation#66317](https://github.com/tgstation/tgstation/pull/66317)
> * [Fixes asset caching tgstation/tgstation#69852](https://github.com/tgstation/tgstation/pull/69852)
> 
> apply at all?
> 
> This will be really helpful for TGUI prefs init time, so I would like this implemented

It was not listed in the PR body, but this PR has these already. I will update.

**Comment by Tsar-Salat** [comment]
> can't be worse.

this means the opposite of what you are trying to say. "Cant/Couldnt be worse" means its rock-bottom and literally cannot be surpassed it is so shitty. 

I dont think Lemon's work is *that* bad.

:trollface:

**Comment by EvilDragonfiend** [comment]
> > can't be worse.
> 
> this means the opposite of what you are trying to say. "Cant/Couldnt be worse" means its rock-bottom and literally cannot be surpassed it is so shitty.
> 
> I dont think Lemon's work is _that_ bad.
> 
> :trollface:

ingulishi difficultu

**Review by Crossedfall — APPROVED** [review]
Tentatively giving my approval. Provided nothing breaks between now and tomorrow's high-pop while this is TM'd it can be merged.

File: code/modules/asset_cache/transports/asset_transport.dm [hunk 7]
```diff
@@ -72,6 +72,7 @@
 		|| asset_cache_item.legacy \
 		|| asset_cache_item.keep_local_name \
 		|| (asset_cache_item.namespace && !asset_cache_item.namespace_parent)
+	// Runtime note: "null.legacy" happens, but that's because a client requested it while the server is still setting up. not that an issue. if it happens, check the comment at "Chat" sprite asset.
```
Comment (itsmeow) [inline code/modules/asset_cache/transports/asset_transport.dm:75]: if a runtime is happening that is "okay", it should be suppressed or blocked from happening.

File: code/modules/tgui/tgui_window.dm [hunk 8]
```diff
@@ -267,8 +298,6 @@
 		return
 	sent_assets |= list(asset)
 	. = asset.send(client)
-	if(!.)
```
Comment (itsmeow) [inline code/modules/tgui/tgui_window.dm:293]: This fixes a bunch of crap. You may want to look at reverting a090357443f9440169ed3f935162116e20e87857 since it appears to have broken a bunch of this.

File: code/modules/tgui/tgui_window.dm [hunk 9]
```diff
@@ -322,24 +363,53 @@
 	// If not locked, handle these message types
 	switch(type)
 		if("ping")
-			send_message("pingReply", payload)
+			send_message("ping/reply", payload)
```
Comment (itsmeow) [inline code/modules/tgui/tgui_window.dm]: this PR is irrelevant and should be atomized out. I would like it ported though.

File: code/modules/tgui/tgui_window.dm [hunk 10]
```diff
@@ -322,24 +363,53 @@
...
-			initialize(
-				fancy = initial_fancy,
-				assets = initial_assets,
-				inline_html = initial_inline_html,
-				inline_js = initial_inline_js,
-				inline_css = initial_inline_css)
-			// Resend the assets
-			for(var/asset in sent_assets)
-				send_asset(asset)
+			reinitialize()
```
Comment (itsmeow) [inline code/modules/tgui/tgui_window.dm]: we already ported these changes in TGUI say

File: code/modules/tgui/tgui_window.dm [hunk 11]
```diff
@@ -322,24 +363,53 @@
...
+		var/with_id = replacetext(command_template, "CONTROL", target_control)
+		var/full_command = replacetext(with_id, "PAYLOAD", event_message)
+
+		var/list/params = list()
+		params["parent"] = "default" //Technically this is external to tgui but whatever
+		params["name"] = mouseMacro
+		params["command"] = full_command
+
+		winset(client, "[mouseMacro]Window[id]Macro", params)
+	mouse_event_macro_set = TRUE
```
Comment (itsmeow) [inline code/modules/tgui/tgui_window.dm]: these macros are irrelevant to us and should not be in this PR

File: code/__HELPERS/icons.dm [hunk 12]
```diff
@@ -1298,12 +1298,13 @@ GLOBAL_LIST_INIT(freon_color_matrix, list("#2E5E69", "#60A2A8", "#A1AFB1", rgb(0
...
 			var/name = SANITIZE_FILENAME("[generate_asset_name(thing)].png")
-			SSassets.transport.register_asset(name, thing)
+			if (!SSassets.cache[name])
+				SSassets.transport.register_asset(name, thing)
 			for (var/thing2 in targets)
 				SSassets.transport.send_assets(thing2, name)
 			if(sourceonly)
 				return SSassets.transport.get_asset_url(name)
-			return "<img class='icon icon-misc' src='[SSassets.transport.get_asset_url(name)]'>"
+			return "<img class='[extra_classes] icon icon-misc' src='[SSassets.transport.get_asset_url(name)]'>"
```
Comment (itsmeow) [inline code/__HELPERS/icons.dm]: what uses this? why is it even being ported?

File: code/modules/asset_cache/asset_list.dm [hunk 13]
```diff
@@ -6,36 +7,59 @@
...
 
-	return cached_url_mappings
+	return cached_serialized_url_mappings
 
 /datum/asset/proc/register()
 	return
 
 /datum/asset/proc/send(client)
-	SHOULD_CALL_PARENT(TRUE)
 	if(!client)
```
Comment (itsmeow) [inline code/modules/asset_cache/asset_list.dm]: we should match TG master on this (see: reverting a090357443f9440169ed3f935162116e20e87857)

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/modules/plumbing/plumbers/pill_press.dm [hunk 14]
```diff
@@ -107,3 +106,4 @@
 			else
 				pill_name = new_name + " pill"
 			. = TRUE
+
```
Comment (itsmeow) [inline code/modules/plumbing/plumbers/pill_press.dm]: line spacing changes in unrelated file

**Review by itsmeow — CHANGES_REQUESTED** [review]
Make job huds cross round cachable

File: code/modules/asset_cache/asset_list.dm (hunk 13 above)
Comment (EvilDragonfiend) [reply code/modules/asset_cache/asset_list.dm]: the asset system will work differently than before, so it would prob fine

Comment (itsmeow) [reply code/modules/asset_cache/asset_list.dm]: I can tell some changes are not in line with TG master, this proc doesn't have the client check on master at all. just removing the should call parent seems bad

**Review by itsmeow — APPROVED** [review]

**Comment by EvilDragonfiend** [comment]
related fixes:
* #9223
* #9222
* some html pages like `View Admin remarks` can cause runtime because html page has image assets but the game didn't load asset yet. It will need a follow-up later fix.
